### ISSUE TITLE
feat(developer, common/lmlayer): predictive modeling for morphologically complex languages

### ISSUE THREAD

**Issue by jahorton** [keymanapp/keyman#3058]
From the Community forums at https://community.software.sil.org/t/keyman-roadmap-march-2020/822/29:

> Being able to identify in the tsv file what types of words can take what types of prefixes and in our language’s case clitic pronouns would allow for prediction of at least suffixes, but also better predictions on verbs, nouns, or whatever should follow.

Also, later in the same post:

> Hunspell has at least been barking up this tree for more than a decade and Fieldworks grammar data can be used to help with word prediction.
>
> Many languages will require specific greetings to be the way you open up a text message. Having those float to the top of suggestions for a new text would be useful. Things like Hola, Assalam Alaykum, Good Evening. These type of message openers could be designated in a TSV file.

**Comment by MayuraVerma** [comment]
Instead of hunspell, please consider nuspell.
It is newer version of hunspell written in c++ and claims three times faster than huspell.

https://github.com/nuspell/nuspell

**Comment by jahorton** [comment]
I just wanted to note that compatibility of some sort with Hunspell dictionaries is probably one of our more common requests.  Upon investigation, it unfortunately looks anything but straightforward, but this is something we may need to tackle due to its popularity at some point.  Just had the topic come up in an email conversation with a user today, and the topic of agglutinative languages - which `hunspell` is specialized for - came up in our recent team planning as well.

The current state of hunspell+JS, so far as I can tell:  there are wrapper libraries for use on `npm` and a few lighter-weight libraries that target various levels of compatibility with hunspell.  The purely-JS libraries seem to be seldomly used and seldomly maintained, which isn't exactly promising.

That said, Hunspell itself _is_ open source (and written in C++, with some reimplementations in other languages) at https://github.com/hunspell/hunspell  It offers MPL licensing, which isn't _quite_ MIT, but it's reasonably permissive and should not impede any attempts to integrate it or convert parts of it if absolutely necessary.  

The file formats of Hunspell dictionaries are plain text, rather than binary.  There are some 'codes' of sorts included in those files; some effort would be required to parse them effectively, but dictionary data is at least reasonably accessible and interpretable.  Also, one of the two backing file types actually somewhat resembles our wordlist .tsv files - though without frequency data.  Here's a link to the data backing a `en` (`en-US`) Hunspell dictionary: https://github.com/wooorm/dictionaries/tree/main/dictionaries/en.  The `.dic` file looks to be a pretty comfortable parse, just with an extra metadata tag, while the `.aff` file... would require some investigation.

**Comment by MayuraVerma** [comment]
Hunspell is older version/legacy

Nuspell is new and faster (3.5x than huspell) 

https://github.com/nuspell/nuspell

Same team has developed Nuspell, if you want to implement Hunspell, please explore Nuspell. It's new library and much faster.

**Comment by mcdurdin** [comment]
Nuspell looks significantly more limited than Hunspell? However, the point is moot because we AFAICT we wouldn't want to use either library internally directly in Keyman, due to licensing, dependency requirements, implementation language (C++ vs web technologies), and not being a precise fit for our needs.

We may consider using the dictionary format, or providing conversion from them to Keyman's dictionary format. Understanding their support for agglutinative languages may help us as well.

**Comment by jahorton** [comment]
Ooh, the `nuspell` wiki offers some great information about the file formats: https://github.com/nuspell/nuspell/wiki#dictionary-maintenance

That, at least, looks to be very useful.

**Comment by jahorton** [comment]
See also #3057 regarding a possible UI idea regarding such models

**Comment by jahorton** [comment]
Examining the `Typo.js` library mentioned above...
1. It's a _limited_ Hunspell implementation.  It doesn't have full support for all possible configurations.
2. It builds a wordlist during the affix + dictionary file parsing phase.  Note the `_applyRule` method in https://github.com/cfinke/Typo.js/blob/00cebf4c839154b55059527fd75b6aacc1598750/ts/typo.ts - that's where affix rules are utilized to expand the wordlist based on stems and affixes.

The benefit of that strategy is that `Typo.js` can be passed raw Hunspell dictionaries and load them... but that also means there's going to be some start-up costs.

The `TRY` definition of affix files is where Hunspell gets its limited notion of probability; it specifies characters from left-to-right in "most frequent" to "least frequent" order.  No actual probability values per character, mind you.  Hunspell appears to assume that a word's frequency is directly correlated to the frequency of each letter comprising the word and uses that to help determine the ordering of its output suggestions.

**Comment by jahorton** [comment]
So, now to use the insights I got from `Typo.js` to see if that illuminates anything in the standard Hunspell or nuspell implementations.

`nuspell` seems far easier for me to parse, and they've got some public documentation that's really useful for finding entry points that will facilitate examining the code.  https://github.com/nuspell/nuspell/wiki/Using-the-software-library indicates how code should integrate with the nuspell library - and thus, the entry points for setting nuspell up and making calls.

What I've been able to parse so far:

This is the core of the affix + dictionary parsing control flow:  https://github.com/nuspell/nuspell/blob/6e46eb31708003fa02796ee8dc0c9e57248ba141/src/nuspell/aff_data.hxx#L165-L173

It appears that https://github.com/nuspell/nuspell/blob/6e46eb31708003fa02796ee8dc0c9e57248ba141/src/nuspell/aff_data.cxx#L1006 (within the `parse_dic` method called in the block linked above) is where words + their flags are placed into the main "hash" data structure for dictionary words.  Unlike with `Typo.js`, `nuspell` does not appear to build affixed word variants at this stage.

Instead, it appears that the `Checker` class there is responsible for handling affixations.  Both `check_simple_word` and `check_word_in_compound` appear to reference a number of `strip_prefix_*` and `strip_suffix_*` methods that appear to be the components it uses for said purpose.  `Checker` itself appears to check for precise text matches, rather than "near matches".

`Checker` is leveraged by `Suggester` to actually make suggestions.  The primary method on the public `Dictionary` type quickly calls `Suggester::suggest_priv`, which appears to handle text casing ('upper', 'lower', 'initial', Pascal + Camel cases).  Its branches appear to leverage the `suggest_low` method, and that appears to be the primary "text correction" method.  It calls a number of helper methods that each seem to leverage the `Checker` or another helper that does.

-----

Based on the patterns I see above... Hunspell looks pretty great for text _correction_, but doesn't really seem built for doing _predictive_ text - its architecture doesn't seem designed to find words that _extend_ the current text - only if the provided data can _match_ the current text token.  Both Hunspell and Nuspell describe themselves as spell _checkers_, after all.

We can probably leverage their definition structures and some of the logic for affixation.  Those seem solid and quite useful... and we could figure out some preprocessing patterns to make things more compatible with our engine.  On the other hand, even if we _could_ plug in either Hunspell or Nuspell wholesale, we _still_ wouldn't get text prediction out of the box.  We'd need to do our own, specialized preprocessing and find a way to re-implement its suggestion paradigm to something far more prediction-compatible in order to have Hunspell-spec based models be able to yield proper predictions.

**Comment by mcdurdin** [comment]
> Based on the patterns I see above... Hunspell and Nuspell look pretty great for text _correction_, but doesn't really seem built for doing _predictive_ text - their architectures don't seem designed to find words that _extend_ the current text - they're designed to _check_ if the provided data can _match_ the current text token. Both Hunspell and Nuspell describe themselves as spell _checkers_, after all.
> 
> We can probably leverage their definition structures and some of the logic for affixation. Those seem solid and quite useful... and we could figure out some preprocessing patterns to make things more compatible with our engine. On the other hand, even if we _could_ plug in either Hunspell or Nuspell wholesale, we _still_ wouldn't get text prediction out of the box. We'd need to do our own, specialized preprocessing and find a way to re-implement its suggestion paradigm to something far more prediction-compatible in order to have Hunspell-spec based models be able to yield proper predictions.

Yes, this was pretty much my assessment, albeit based on a much less in-depth review of the code. We may be able to make use of some of their data structures or raw data but will need to rewrite the logic to fit our requirements. Tooling to import a Hunspell dictionary would also be good -- even for morphologically straightforward languages.

**Comment by jahorton** [comment]
> > [..]
> 
> Yes, this was pretty much my assessment, albeit based on a much less in-depth review of the code. We may be able to make use of some of their data structures or raw data but will need to rewrite the logic to fit our requirements. Tooling to import a Hunspell dictionary would also be good -- even for morphologically straightforward languages.

I believe it would be possible to _fully utilize_ their raw data for affix + dictionary specifications.  It'd take some effort, and would likely be best done by starting small and progressively adding support for more leveraged features.  We'd also certainly want to precompile it into something more runtime-friendly for our engine, though.  I think we can also leverage some of their logic; if nothing else, inspecting their code is quite helpful for understanding their spellchecking model and thinking through its implications.
